src/eval.rs

```diff
@@ -1,26 +1,998 @@
+                    BOp::Mul => cast(l * r, l_range, Overflow::Wraparound),
+                    BOp::Div => {
+                        if r == 0 {
+                            Err(RuntimeError::DivisionbyZero)
+                        } else {
+                            cast(l / r, l_range, Overflow::Wraparound)
+                        }
+                    }
+                    BOp::Rem => {
+                        if r == 0 {
+                            Err(RuntimeError::ModulusByZero)
+                        } else {
+                            cast(l % r, l_range, Overflow::Wraparound)
+                        }
+                    }
+                    BOp::Shl => cast(l << r, l_range, Overflow::Wraparound),
+                    BOp::Shr => cast(l >> r, l_range, Overflow::Wraparound),
+
+                    // boolean return
+                    BOp::Lt => Ok(Value::Bool(l < r)),
+                    BOp::Lte => Ok(Value::Bool(l <= r)),
+                    BOp::Eq => Ok(Value::Bool(l == r)),
+                    BOp::Ne => Ok(Value::Bool(l != r)),
+
+                    _ => Err(RuntimeError::InvalidOperand),
+                },
+
+                (Value::Bool(l), Value::Bool(r)) => match op {
+                    BOp::And => Ok(Value::Bool(l && r)),
+                    BOp::Or => Ok(Value::Bool(l || r)),
+
+                    _ => Err(RuntimeError::InvalidOperand),
+                },
+                _ => Err(RuntimeError::TypeError),
+            }
+        }
+
+        Expr::UOp { op, inner } => {
+            let left = eval_expr(inner, env, &program)?;
+
+            match left {
+                Value::Num(l, range) => match op {
+                    UOp::Negate => cast(-l, range, Overflow::Wraparound),
+                    _ => Err(RuntimeError::InvalidOperand),
+                },
+
+                Value::Bool(b) => match op {
+                    UOp::Not => Ok(Value::Bool(!b)),
+                    _ => Err(RuntimeError::InvalidOperand),
+                },
+                _ => Err(RuntimeError::TypeError),
+            }
+        }
+
+        Expr::Cast {
+            inner,
+            typ,
+            overflow,
+        } => {
+            let cast_t = match typ {
+                Type::BoolT => Err(RuntimeError::TypeError),
+                Type::SymT => Err(RuntimeError::TypeError),
+                Type::NumT(n) => Ok(n.clone()),
+            }?;
+
+            let val = eval_expr(inner, env, &program)?;
+
+            match val {
+                Value::Num(num, range) => match overflow {
+                    Overflow::Fail => cast(num, cast_t, Overflow::Fail),
+                    Overflow::Saturate => cast(num, cast_t, Overflow::Saturate),
+                    Overflow::Wraparound => cast(num, cast_t, Overflow::Wraparound),
+                },
+                // can only cast ints
+                Value::Bool(b) => Err(RuntimeError::TypeError),
```

Comment:
merge this and the next case (e.g. `_ => Err(...)`). Also, do not name unused variables (use `_`). do this everywhere else.